src/EventStore.Core/ProcessTracker.cs

```diff
@@ -0,0 +1,190 @@
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Threading;
+using EventStore.Common.Utils;
+using EventStore.Core.Services.Monitoring.Stats;
+using EventStore.Core.Services.Monitoring.Utils;
+using EventStore.Core.Time;
+using EventStore.Native.Monitoring;
+using HostStat;
+using Serilog;
+
+namespace EventStore.Core;
+
+public class ProcessTracker {
+	private static readonly ILogger _log =
+		Serilog.Log.ForContext(Serilog.Core.Constants.SourceContextPropertyName, "PROCESS-TRACKER-LOGGER");
+
+	private IClock _clock;
+	private Instant _lastTime = Instant.FromSeconds(0);
+	private Process? _process;
+	private LoadAverages _loadAverages;
+	private EsDriveInfo? _esDriveInfo;
+	private DiskIo? _diskIo;
+
+
+	private readonly HostStat.HostStat _stats;
+	private readonly DateTime _startTime;
+	private readonly TimeSpan _startCpuUsage;
+	private readonly string _dbPath;
+	private readonly PerfCounterHelper _perfCounter;
+	private readonly EventCountersHelper _eventCountersHelper;
+
+	public ProcessTracker(IClock clock, string dbPath) {
+		_clock = clock;
+		_stats = new HostStat.HostStat();
+		_dbPath = dbPath;
+		_perfCounter = new PerfCounterHelper(_log);
+		_startTime = DateTime.Now;
+		_startCpuUsage = Process.GetCurrentProcess().TotalProcessorTime;
+		_eventCountersHelper = new EventCountersHelper(2_500);
+	}
+
+	private T Bookkeeping<T>(Func<T> func) {
+		var now = _clock.Now;
+		if (_lastTime != Instant.FromSeconds(0) && now.ElapsedSecondsSince(_lastTime) < 5)
+			return func();
+
+		_lastTime = now;
+
+		if (OS.IsUnix) {
+			_loadAverages = _stats.GetLoadAverages();
+		}
+
+		_esDriveInfo = EsDriveInfo.FromDirectory(_dbPath, _log);
+		_diskIo = DiskIo.GetDiskIo(Environment.ProcessId, _log);
+		_process = Process.GetCurrentProcess();
+
+		return func();
+	}
+
+	public double CollectCpuAvg1M() {
+		return Bookkeeping(() => _loadAverages.Average1m);
+	}
+
+	public double CollectCpuAvg5M() {
+		return Bookkeeping(() => _loadAverages.Average5m);
+	}
+
+	public double CollectCpuAvg15M() {
+		return Bookkeeping(() => _loadAverages.Average15m);
+	}
+
+	public float CollectCpuTotalUsage() {
+		return Bookkeeping(() => _perfCounter.GetTotalCpuUsage());
+	}
+
+	public long CollectFreeMemoryUnix() {
+		return Bookkeeping(() => (long)_stats.GetFreeMemory());
+	}
+
+	public long CollectFreeMemoryWindows() {
+		return Bookkeeping(() => _perfCounter.GetFreeMemory());
+	}
+
+	public long CollectTotalMemoryUnix() {
+		return Bookkeeping(() => (long)_stats.GetTotalMemory());
+	}
+
+	public long CollectTotalMemoryWindows() {
+		return Bookkeeping(() => (long)WinNativeMemoryStatus.GetTotalMemory());
+	}
+
+	public Measurement<double> CollectProcCPU() {
+		return Bookkeeping(() => {
+			var dim = new KeyValuePair<string, object?>[] {
+				new("id", _process!.Id),
+				new("mem", _process!.WorkingSet64),
+				new("startTime", _process!.StartTime.ToUniversalTime().ToString("O")),
+				new("threadsCount", ThreadPool.ThreadCount),
```

Comment:
ah, so, dimensions dont really work like this. each tuple in the cartestian product of the dimensions is a separate timeseries. so here if the threadscount changes we would start reporting the procpuusage on a different timeseries

Reply:
Addressed